docs: explain stream error handling (#5746)
fastify/fastify: docs/Reference/Server.md

  sent to the client. Use the `onSend` hook instead.
- not found (404) errors. Use [`setNotFoundHandler`](#set-not-found-handler)
  instead.

> - stream errors thrown during pipe-ing process into the response socket as the
>   headers/response was already sent to the client. Use custom in-stream data
>   to signal such errors to client.

Reviewer: ```suggestion
- Stream errors thrown during piping into the response socket, as headers/response 
  were already sent to the client. Use custom in-stream data to signal such errors.
```

  log.error(error)
}
```

Custom error handler has to account for stream replies.

If `res.send(jsonData)` is used in the error handler, it will fail for the stream
response because Fastify will not be able to serialize it.
Always using string data in error handler replies will avoid this issue.
For example, manually serializing the error response to JSON

> with `JSON.stringify`.

Reviewer 2: I'll leverage this to @Fdawgs. I believe this needs rewording, I found it a bit misguiding; couldn't fully understand what the problem was without seeing the tests

  log.error(error)
}
```

When you are sending streams you will have to account for that in the custom
error handler.

> If you use `res.send(jsonData)` in the error handler it will fail for stream

Reviewer 3: We don't use "you" in the docs

Example:

```suggestion
If `res.send(jsonData)` is used in the error handler, it will fail for stream
```

Author: Oh, I referenced some docs in Reply that use `you`. I'll update.

Author: updated

Reviewer 3: Yeah, we're trying to slowly unify everything under one style

There are some inconsistencies at the moment, and we're not fully there...

  log.error(error)
}
```

> Custom error handler has to account for stream replies.

Reviewer 4: ```suggestion
### Custom error handler has to account for stream replies.
```

> ### Custom error handler has to account for stream replies.
> 
> If `Content-Type` is different between endpoint and error handler
> For example, endpoint returns `application/text` stream and error handler
> responds with `application/json` JSON data. In this case `Content-Type` has
> to be explicitly defined in both endpoint and error handler. Otherwise, the
> error handler would fail serialization with `500` status code as it would not
> apply JSON serialization logic.
> Another option would be to always respond with serialized data in error handler,
> by manually calling serialization method (e.g. `JSON.stringify`).

Reviewer: ```suggestion
##### Custom error handler has to account for stream replies

If `Content-Type` differs between the endpoint and error handler, explicitly
define it in both. For example, if the endpoint returns an `application/text`
stream and the error handler responds with `application/json`, the error handler
must explicitly set `Content-Type`. Otherwise, it will fail serialization with
a `500` status code. Alternatively, always respond with serialized data in the
error handler by manually calling a serialization method (e.g.,
`JSON.stringify`).
```
